trieste/utils/objectives.py

```diff
@@ -126,6 +128,143 @@ def logarithmic_goldstein_price(x: TensorType) -> TensorType:
+        """
+        Generate `n` pareto optimal points
+        """
+
+
+class VLMOP2(MultiObjectiveTestProblem):
+    """
+    The VLMOP2 function, typically evaluated over :math:`[-2, 2]^2`. See
+    :cite:`van1999multiobjective`  for details.
+    """
+
+    bounds = [[-2.0] * 2, [2.0] * 2]
+    dim = 2
+
+    def prepare_benchmark(self):
+        return vlmop2
+
+    def gen_pareto_optimal_points(self, n: int):
+        """
+        :cite: fonseca1995multiobjective
+        True pareto lies on -1/sqrt(2)-1/sqrt(2) and x1=x2
+        """
+        tf.debugging.assert_greater(n, 0)
+        _x = tf.linspace([-1 / tf.sqrt(2.0)], [1 / tf.sqrt(2.0)], n)
+        return vlmop2(tf.concat([_x, _x], axis=1))
+
+
+def vlmop2(x: TensorType) -> TensorType:
+    tf.debugging.assert_equal(x.shape[-1], 2, "vlmop2 only allow 2d input")
+    transl = 1 / tf.sqrt(2.0)
+    y1 = 1 - tf.exp(-1 * tf.reduce_sum((x - transl) ** 2, axis=1))
+    y2 = 1 - tf.exp(-1 * tf.reduce_sum((x + transl) ** 2, axis=1))
+    return tf.stack([y1, y2], axis=1)
+
+
+class DTLZ(MultiObjectiveTestProblem):
+    """
+    DTLZ series multi-objective test functions.
+    :cite: deb2002scalable
+    """
+
+    def __init__(self, input_dim: int, num_objective: int):
+        tf.debugging.assert_greater(input_dim, 0)
+        tf.debugging.assert_greater(num_objective, 0)
+        tf.debugging.assert_greater(
+            input_dim,
+            num_objective,
+            f"input dimension {input_dim}"
+            f"  must be greater than function objective numbers {num_objective}",
+        )
+        self.dim = input_dim
+        self.M = num_objective
+        self.k = self.dim - self.M + 1
+        self.bounds = [[0] * input_dim, [1] * input_dim]
+        super().__init__()
+
+
+class DTLZ1(DTLZ):
+    def prepare_benchmark(self):
+        return partial(dtlz1, m=self.M, k=self.k, d=self.dim)
+
+    def gen_pareto_optimal_points(self, n: int, seed=None):
+        tf.debugging.assert_greater_equal(self.M, 2)
+        rnd = tf.random.uniform([n, self.M - 1], minval=0, maxval=1)
+        strnd = tf.sort(rnd, axis=-1)
+        strnd = tf.concat([tf.zeros([n, 1]), strnd, tf.ones([n, 1])], axis=-1)
+        return 0.5 * (strnd[..., 1:] - strnd[..., :-1])
+
+
+def dtlz1(x: TensorType, m: int, k: int, d: int) -> TensorType:
+    tf.debugging.assert_equal(
+        x.shape[-1], d, f"input x dim: {x.shape[-1]} is not align with pre-specified dim: {d}"
+    )
+
+    def g(xM):
+        return 100 * (
+            k
+            + tf.reduce_sum(
+                (xM - 0.5) ** 2 - tf.cos(20 * math.pi * (xM - 0.5)), axis=-1, keepdims=True
+            )
+        )
+
+    f = None
+    for i in range(m):
+        xM = x[:, m - 1 :]
+        y = 1 + g(xM)
+        y *= 1 / 2 * tf.reduce_prod(x[:, : m - 1 - i], axis=1, keepdims=True)
+        if i > 0:
+            y *= 1 - x[:, m - i - 1, tf.newaxis]
+        f = y if f is None else tf.concat([f, y], 1)
+    return f
```

Comment:
it would be good if these functions were compatible with tf.function. In particular, `tf.function` doesn't like local mutability with `tf.Tensor`. Would it be difficult to do that? You could use broadcasting or `tf.Variable`

Reply:
Seems I didn't see a direct way of using broadcasting here, correct me if I missed anything as I was not very familiar with checking the compatibility of `tf.function`. 

Alternatively, is it a good idea of using `tf.map_fn` here as below? 
```
def dtlz1_tffver(x: TensorType, m: int, k: int, d: int) -> TensorType:
    tf.debugging.assert_equal(
        x.shape[-1], d, f"input x dim: {x.shape[-1]} is not align with pre-specified dim: {d}"
    )

    def g(xM):
        return 100 * (
            k
            + tf.reduce_sum(
                (xM - 0.5) ** 2 - tf.cos(20 * math.pi * (xM - 0.5)), axis=-1, keepdims=True
            )
        )

    def fm(input):
        x, i = input
        y = 1 + g(x[:, m - 1:])
        y *= 1 / 2 * tf.reduce_prod(x[:, : m - 1 - i], axis=1, keepdims=True)
        if i > 0:
            y *= 1 - x[:, m - i - 1, tf.newaxis]
        return y

    test = tf.map_fn(fm, (tf.tile(x[tf.newaxis, ...], [m, 1, 1]),
                          tf.range(m, dtype=tf.int32)), dtype=(x.dtype, tf.int32),
                          fn_output_signature=x.dtype)
    return tf.concat(tf.unstack(test), 1)
```

Reply:
Was this resolved?

Reply:
Nope, I was not aware of an easy way to use broadcasting here, the only alternative way came in my mind is something as above, still expect some suggestions here.

Reply:
Perhaps @uri-granta has some suggestions? I'm afraid I don't.

 I'm not sure that it matters if the synthetic objective functions are not fully optimized w.r.t speed. Any real-world objective functions certainly will not be. Kind of the point of BO!

Reply:
I'm still new to tensorflow, but wouldn't something like this, using `tf.TensorArray`, make it compatible with tf.function?

```python
def dtlz2(x: TensorType, m: int, d: int) -> TensorType:

    def g(xM):
        z = (xM - 0.5) ** 2
        return tf.reduce_sum(z, axis=1, keepdims=True)

    ta = tf.TensorArray(x.dtype, size=m)
    for i in tf.range(m):
        y = 1 + g(x[:, m - 1 :])
        for j in tf.range(m - 1 - i):
            y *= tf.cos(math.pi / 2 * x[:, j, tf.newaxis])
        if i > 0:
            y *= tf.sin(math.pi / 2 * x[:, m - 1 - i, tf.newaxis])
        ta = ta.write(i, y)
    f = some_sort_of_transposition_to_make_the_right_shape(ta.stack())
    return f
```